### PR TITLE
[5.4] Add @errors blade directive

### PR DESCRIPTION
**Pull request by m1guelpf**
This Pull Requests adds an @errors blade directive.

Original idea by @paulredmond in Twitter.

## Structure

I feel like the oneliner of the implementation is very big. Would it be better to create a errors() helper and move the logic there?
ping @themsaid @taylorotwell

### CODE REVIEW

**Comment by browner12** [comment]
i think this belongs in userland

**Comment by m1guelpf** [comment]
@browner12 I think every CRUD application could benefit from this at least one time.

**Comment by devcircus** [comment]
Of course, but there's no end to these helpers that could be beneficial and they're very simple to implement.

**Comment by m1guelpf** [comment]
@devcircus Isn't Laravel's job to make developer's lifes easier?

**Comment by browner12** [comment]
yah, just seems like a slippery slope here

**Comment by paulredmond** [comment]
I appreciate the sentiment, but I agree the original idea was in my own projects.

I'm also not sure how to get at the view variables and I'd cache this with static var or something so t doesn't hit the session store every time.

**Comment by m1guelpf** [comment]
@paulredmond That's why I was suggesting to move the logic to a helper. Will do it tomorrow if it keeps open

**Comment by paulredmond** [comment]
Every time something is added to core it must be maintained indefinitely is a non-breaking release. That list can add up quick, especially around small view helpers. I hope you don't take it personally as bad code or a bad PR 💖

**Comment by m1guelpf** [comment]
@paulredmond I understand, but the amount of times I've had to write that manually makes me think it'd be awesome to have it on core... Let's wait for @taylorotwell's review...

**Comment by taylorotwell** [comment]
Before we talk about if it can be included can you please clean-up that ternary expression. I can't even read the code tbh. Thanks.

**Comment by vlakoff** [comment]
This directive doesn't sound clear to me, it sounds more like an iterable list than a conditional. I prefer the current syntax.

**Comment by fernandobandeira** [comment]
Another thing is that we can change the error variable name if we have more than one form on the same page, this won't cover this usecase...

**Comment by m1guelpf** [comment]
@fernandobandeira What do you mean? I don't understand...

**Comment by m1guelpf** [comment]
@taylorotwell Cleaned up. If you approve this, I can add tests :smile:

**Comment by fernandobandeira** [comment]
@m1guelpf 

```php
return redirect('register')
            ->withErrors($validator, 'login');

{{ $errors->login->first('email') }}
```

There's no way to filter the error bag using this new directive.

**Comment by m1guelpf** [comment]
@fernandobandeira Will work in this tomorrow (if this is still open...)

**Comment by deleugpn** [comment]
What about error bags that differs from `default`?

**Comment by m1guelpf** [comment]
@deleugpn See @fernandobandeira's comment above and my answer :smile:

**Comment by taylorotwell** [comment]
I'm not super opposed to a short-cut for this syntax as I agree it is used a lot in Laravel. There are some things that would need to be addressed (non-default bag, etc.) as others have mentioned. Will keep it open for now as I'm not totally opposed to the idea.

**Comment by vlakoff** [comment]
Of course adding an helper `errors()` would be misleading and cluttering. It doesn't return errors but a boolean, it should be named `has_errors()`. Even better, it shouldn't be added. Laravel helpers are already occupying so many generic words.

**Comment by m1guelpf** [comment]
@vlakoff It's better than having the logic in blade... I will rename it to has_errors, though. Do you have any other ideas about where the logic could be?

**Comment by vlakoff** [comment]
I would suggest this:
```php
    protected function compileErrors($key = null)
    {
        if (is_null($key)) {
            return "<?php if(session('errors') && session('errors')->any()): ?>";
        }
    
        return "<?php if(session('errors') && session('errors')->has({$key})): ?>";
    }
```
<br>By the way, you could have false positives if `session('errors')` is not null, but an empty MessageBag.

**Comment by vlakoff** [comment]
And a slightly optimized alternative:

```php
    protected function compileErrors($key = null)
    {
        if (is_null($key)) {
            return "<?php if(($errors = session('errors')) && $errors->any()): ?>";
        }

        return "<?php if(($errors = session('errors')) && $errors->has({$key})): ?>";
    }
```

**Comment by vlakoff** [comment]
Please disregard the optimized alternative. We don't want to pollute the view by adding/altering variables.

**Comment by m1guelpf** [comment]
@vlakoff The problem is that the logic will be more complex than that when I add custom non-default bags, and also Blade caches aggresively. I think I'll keep the helper, and we'll see what Taylor thinks.

**Comment by m1guelpf** [comment]
I've renamed the helper, added support for custom error bags & fixed a false positive. This is now ready for review. If @taylorotwell approves this, I'll add tests.

File: src/Illuminate/View/Compilers/Concerns/CompilesConditionals.php [hunk 1]
```diff
@@ -119,4 +119,25 @@ protected function compileEndAuth()
...
+
+    /**
+     * Compile the if-errors statements into valid PHP.
+     *
+     * @param  string|null  $key
+     * @return string
+     */
+    protected function compileErrors($key = null)
+    {
+        return "<?php if(errors({$key})): ?>";
```
Comment (fernandobandeira) [inline src/Illuminate/View/Compilers/Concerns/CompilesConditionals.php]: This is still using the old ``` errors ``` helper, also we can't pass ``` $bag ``` I think you could change ``` $key ``` to accept an array too. This way:

```

@errors
Has errors
@enderrors

@errors('name')
Name field has errors
@enderrors

@errors(['bag' => 'login'])
Login form has errors
@enderrors

@errors(['bag' => 'login', 'key' => 'name'])
Name field of login form has errors
@enderrors

File: src/Illuminate/Foundation/helpers.php [hunk 2]
```diff
@@ -432,6 +432,27 @@ function encrypt($value)
...
+     * Returns true if session has errors, and optionally key-specific errors.
+     *
+     * @param  string|null  $key
+     * @param  string|null  $bag
+     * @return bool
+     */
+    function has_errors($key = null, $bag = null)
+    {
+        $errors = session('errors');
+        if (is_null($errors) && ! session('errors')->any()) {
```
Comment (vlakoff) [inline src/Illuminate/Foundation/helpers.php]: Logic is wrong. It should be `if (is_null($errors) || ! $errors->any())`

File: src/Illuminate/Foundation/helpers.php [hunk 3]
```diff
@@ -432,6 +432,27 @@ function encrypt($value)
...
+     * Returns true if session has errors, and optionally key-specific errors.
+     *
+     * @param  string|null  $key
+     * @param  string|null  $bag
+     * @return bool
+     */
+    function has_errors($key = null, $bag = null)
+    {
+        $errors = session('errors');
+        if (if (is_null($errors) || ! $errors->any())) {
```
Comment (vlakoff) [inline src/Illuminate/Foundation/helpers.php]: Please put the bottle down

File: src/Illuminate/Foundation/helpers.php [hunk 4]
```diff
@@ -443,7 +443,7 @@ function encrypt($value)
     function has_errors($key = null, $bag = null)
     {
         $errors = session('errors');
-        if (if (is_null($errors) || ! $errors->any())) {
+        if (is_null($errors) || ! $errors->any()) {
```
Comment (vlakoff) [inline src/Illuminate/Foundation/helpers.php]: Also it always checks the default bag, even if a bag has been specified.

Example of fix:
`is_null($errors) || ! $errors->getBag($bag ?: 'default')->any()`

Comment (vlakoff) [inline src/Illuminate/Foundation/helpers.php]: Even better, at top of code: `$bag = $bag ?: 'default'`

Then for the rest of the code you can do: `$errors->$bag->...`

**Comment by taylorotwell** [comment]
Please provide code examples of using all features, including custom bags, etc.

**Comment by m1guelpf** [comment]
@taylorotwell There you have!
errorhelper

File: src/Illuminate/View/Compilers/Concerns/CompilesConditionals.php [hunk 5]
```diff
@@ -140,4 +140,25 @@ protected function compileEndIsset()
...
+
+    /**
+     * Compile the if-errors statements into valid PHP.
+     *
+     * @param  string|null  $key
+     * @return string
+     */
+    protected function compileErrors($expression = null)
+    {
+        return "<?php if(errors{$expression}): ?>";
```
Comment (fernandobandeira) [inline src/Illuminate/View/Compilers/Concerns/CompilesConditionals.php]: Shouldn't it be ``` has_errors ``` ?
Also the docblock isn't matching the parameter name anymore
```
@param  string|null  $key
```

**Comment by deleugpn** [comment]
@m1guelpf could you please try 

`@includeWhen(@errors('field'), 'my-error-field-component')`

File: src/Illuminate/Foundation/helpers.php [hunk 6]
```diff
@@ -470,6 +470,27 @@ function factory()
...
 
+if (! function_exists('has_errors')) {
+    /**
+     * Returns true if session has errors, and optionally key-specific errors.
+     *
+     * @param  string|null  $key
+     * @param  string|null  $bag
+     * @return bool
+     */
+    function has_errors($key = null, $bag = null)
```
Comment (deleugpn) [inline src/Illuminate/Foundation/helpers.php]: `$bag = 'default'`.

`$errors->getBag($bag)->any()`

File: src/Illuminate/Foundation/helpers.php [hunk 7]
```diff
@@ -470,6 +470,27 @@ function factory()
...
+    {
+        $errors = session('errors');
+        if (is_null($errors) || ! $errors->getBag($bag ?: 'default')->any()) {
+            return false;
+        } elseif ($key !== null) {
+            return is_null($bag) ? $errors->has($key) : $errors->$bag->has($key);
```
Comment (deleugpn) [inline src/Illuminate/Foundation/helpers.php]: `$errors->getBag($bag)->has($key);`

File: src/Illuminate/Foundation/helpers.php [hunk 8]
```diff
@@ -470,6 +470,27 @@ function factory()
...
+     * Returns true if session has errors, and optionally key-specific errors.
+     *
+     * @param  string|null  $key
+     * @param  string|null  $bag
+     * @return bool
+     */
+    function has_errors($key = null, $bag = null)
+    {
+        $errors = session('errors');
+        if (is_null($errors) || ! $errors->getBag($bag ?: 'default')->any()) {
```
Comment (deleugpn) [inline src/Illuminate/Foundation/helpers.php]: `$errors->getBag($bag)->any()`

**Review by fernandobandeira — APPROVED** [review]
Seems way better 😄

File: src/Illuminate/Foundation/helpers.php [hunk 9]
```diff
@@ -470,6 +470,27 @@ function factory()
...
 
+if (! function_exists('has_errors')) {
+    /**
+     * Returns true if session has errors, and optionally key-specific errors.
+     *
+     * @param  string|null  $key
+     * @param  string|null  $bag
+     * @return bool
+     */
+    function has_errors($key = null, $bag = 'default')
```
Comment (vlakoff) [inline src/Illuminate/Foundation/helpers.php]: Should be changed back to `$bag = null`, then add something like `$bag = $bag ?: 'default'` at beginning of code. One should be able to programmatically use the default bag by passing `null`.

File: src/Illuminate/Foundation/helpers.php [hunk 10]
```diff
@@ -470,6 +470,27 @@ function factory()
...
+    {
+        $errors = session('errors');
+        if (is_null($errors) || ! $errors->getBag($bag)->any()) {
+            return false;
+        } elseif ($key !== null) {
+            return $errors->$bag->has($key);
```
Comment (vlakoff) [inline src/Illuminate/Foundation/helpers.php]: Pick `->getBag($bag)` or `->$bag`, but not both.

File: src/Illuminate/Foundation/helpers.php (hunk 6 above)
Comment (deleugpn) [inline src/Illuminate/Foundation/helpers.php]: `Null` is inconsistent with other parts of the framework. I think it should stay as `default` https://github.com/laravel/framework/blob/5.4/src/Illuminate/Support/ViewErrorBag.php#L23

File: src/Illuminate/Foundation/helpers.php (hunk 9 above)
Comment (deleugpn) [inline src/Illuminate/Foundation/helpers.php]: I disagree with this statement, but even if I agreed to it, it would have to be a separate PR to also change https://github.com/laravel/framework/blob/5.4/src/Illuminate/Support/ViewErrorBag.php#L23 for consistency.

Comment (vlakoff) [inline src/Illuminate/Foundation/helpers.php]: Thanks for pointing out this, indeed it should be consistent.

It's pretty standard to pass `null` for using the default value, so one doesn't have to remember if it's "default", "\_default", "\_\_default\_\_" or whatever. But indeed this is a different changeset.

Comment (m1guelpf) [inline src/Illuminate/Foundation/helpers.php]: @vlakoff I think I'll revert back to 'default' for consistency. You can submit a PR changing both if this gets merged.

Comment (m1guelpf) [inline src/Illuminate/Foundation/helpers.php]: @vlakoff Anyway, if you don't specify the second value, it'll assume 'default'. You don't need to remember it :smile:

**Review by deleugpn — APPROVED** [review]

**Comment by m1guelpf** [comment]
@taylorotwell Everything is ready for review now :octocat:

File: src/Illuminate/Foundation/helpers.php (hunk 9 above)
Comment (vlakoff) [inline src/Illuminate/Foundation/helpers.php]: I said "programmatically", i.e. there are some cases you are forcibly specifying an argument.

**Comment by taylorotwell** [comment]
I don't like being forced to pass null in order to check non-default bag. Feels clunky.

**Comment by alexbowers** [comment]
Could it not be done as part of dot syntax?

`@errors('<bag>.array.syntax')` 

Checking against the array syntax first for default array dot syntax, and then falling back to using the first option as the name of a bag, if no value is found.

**Comment by deleugpn** [comment]
Can people use `.` as the name of an input? If that doesn't get translated to array, then @alexbowers would not be feasible.
